core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderBuilder.scala

```diff
@@ -1,22 +1,90 @@
 package org.broadinstitute.dsde.rawls.entities.datarepo
 
-import org.broadinstitute.dsde.rawls.dataaccess.{GoogleBigQueryServiceFactory, SamDAO}
+import java.util.UUID
+
+import akka.http.scaladsl.model.StatusCodes
+import bio.terra.datarepo.client.{ApiException => DatarepoApiException}
+import bio.terra.workspace.client.{ApiException => WorkspaceApiException}
+import bio.terra.workspace.model.DataReferenceDescription.ReferenceTypeEnum
+import com.typesafe.scalalogging.LazyLogging
 import org.broadinstitute.dsde.rawls.dataaccess.datarepo.DataRepoDAO
 import org.broadinstitute.dsde.rawls.dataaccess.workspacemanager.WorkspaceManagerDAO
+import org.broadinstitute.dsde.rawls.dataaccess.{GoogleBigQueryServiceFactory, SamDAO}
 import org.broadinstitute.dsde.rawls.entities.EntityRequestArguments
 import org.broadinstitute.dsde.rawls.entities.base.EntityProviderBuilder
+import org.broadinstitute.dsde.rawls.entities.exceptions.DataEntityException
+import org.broadinstitute.dsde.rawls.model.DataReferenceModelJsonSupport.TerraDataRepoSnapshotRequestFormat
+import org.broadinstitute.dsde.rawls.model.{DataReferenceName, TerraDataRepoSnapshotRequest}
+import spray.json._
 
 import scala.concurrent.ExecutionContext
 import scala.reflect.runtime.universe._
+import scala.util.{Failure, Success, Try}
 
 class DataRepoEntityProviderBuilder(workspaceManagerDAO: WorkspaceManagerDAO, dataRepoDAO: DataRepoDAO,
                                        samDAO: SamDAO, bqServiceFactory: GoogleBigQueryServiceFactory)
                                    (implicit protected val executionContext: ExecutionContext)
-  extends EntityProviderBuilder[DataRepoEntityProvider] {
+  extends EntityProviderBuilder[DataRepoEntityProvider] with LazyLogging {
 
   override def builds: TypeTag[DataRepoEntityProvider] = typeTag[DataRepoEntityProvider]
 
-  override def build(requestArguments: EntityRequestArguments): DataRepoEntityProvider = {
-    new DataRepoEntityProvider(requestArguments, workspaceManagerDAO, dataRepoDAO, samDAO, bqServiceFactory)
+  override def build(requestArguments: EntityRequestArguments): Try[DataRepoEntityProvider] = {
+    for {
+      dataReferenceName <- requestArguments.dataReference.toRight(new DataEntityException("data reference must be defined for this provider")).toTry
+
+      // get snapshot UUID from data reference name
+      snapshotId <- Try(lookupSnapshotForName(dataReferenceName, requestArguments))
+
+      // contact TDR to describe the snapshot
+      snapshotModel <- Try(dataRepoDAO.getSnapshot(snapshotId, requestArguments.userInfo.accessToken)).recoverWith {
+        case notFound: DatarepoApiException if notFound.getCode == StatusCodes.NotFound.intValue =>
+          Failure(new DataEntityException(s"Snapshot id $snapshotId does not exist or you do not have access", notFound))
```

Comment:
Message is confusing given the next case that actually checks if access was denied.

Reply:
no access can also look like it does not exist